colbert/tests/index_updater_test.py

```diff
@@ -0,0 +1,186 @@
+    question = squad_dev[10].question
+    all_tests_passed = True
+    
+# Search full index
+    results = searcher.search(question, k=k)
+    top_k_ids = []
+    for passage_id, passage_rank, passage_score in zip(*results):
+        top_k_ids.append(passage_id)
+    print(f'Initial search top results: {top_k_ids}')
+        
+# Initialize IndexUpdater
+    config = ColBERTConfig(
+        doc_maxlen=doc_maxlen,
+        nbits=nbits,
+        root=experiment_dir,
+        experiment=experiment,
+    )
+    
+    index_updater = IndexUpdater(config, searcher, checkpoint)
+    
+# Remove the first passage from top-k results (no persisting to disk)
+    n = 1
+    index_updater.remove(top_k_ids[:n])
+    
+# Search again without reloading the searcher
+    results = searcher.search(question, k=k)
+    top_k_ids_after = []
+    for passage_id, passage_rank, passage_score in zip(*results):
+        top_k_ids_after.append(passage_id)
+
+# Test that top passage is removed
+    print(f'Top results after removing the top passage: {top_k_ids_after}')
+    if top_k_ids[n:] == top_k_ids_after[:-n] and top_k_ids_after[-n:] != top_k_ids[:n]:
+        print("REMOVAL SUCCEEDED")
+    else:
+        print(top_k_ids)
+        print(top_k_ids_after)
+        
+        print("REMOVAL FAILED!!!")
+        all_tests_passed = False
+
+        
+# Add the removed passage back
+    passage_removed = "Diego on May 24, 1984 during their May 23–25, 1984 meetings in Washington, D.C. This was the first Super Bowl to be played at Jack Murphy Stadium (now currently known as Qualcomm Stadium) in San Diego, California. Fourteen cities were part of the bidding process, which was scheduled to award four Super Bowls (XXI, XXII, XXIII, and XXIV). The bidding cities included: Anaheim, Detroit, Houston, Jacksonville, Miami, Minneapolis, New Orleans, Pasadena, Philadelphia, San Francisco, San Diego, Seattle, Tampa, and Tempe. The Philadelphia host committee assembled what was considered a strong, but long-shot bid, hoping to win the first outdoor Super"   
+    new_pids = index_updater.add([passage_removed])
+    
+# Search again without reloading the searcher
+    results = searcher.search(question, k=k)
+    top_k_ids_after_append = []
+    for passage_id, passage_rank, passage_score in zip(*results):
+        top_k_ids_after_append.append(passage_id)
+        
+    print(f'Top results after appending back the top passage: {top_k_ids_after_append}')
+    
+    if top_k_ids[1:] == top_k_ids_after_append[1:] and top_k_ids_after_append[0] == new_pids[0]:
+        print("REAPPEND SUCCEEDED")
+    else:
+        print("REAPPEND FAILED!!!")
+        all_tests_passed = False
+    
+        
+# Add 4 more copies of the top passages  
+    new_pids.extend(index_updater.add([passage_removed, passage_removed, passage_removed, passage_removed]))
+    
+# Search again without reloading the searcher
+# Now we expect the new 5 identical ids of the top passage to be returned as seearch results
```

Comment:
Are the pids identical? They are different pids but the passage text is the same. Right?

Reply:
Yes, the passages text are identical but pids should be consecutive -- I'll modify this comment to make it more clear.